### ISSUE TITLE
Allow python paths passed via --python-path to overwrite default search path

### ISSUE THREAD

**Issue by tkdchen** [GrahamDumpleton/mod_wsgi#70]
It would be much useful to allow custom `--python-path`s to overwrite default search path, whatever a system-wide Python interpreter is used or a virtual environment is used with or without `--system-package-sites`. Currently, in the `master` and `develop` branch, custom python paths are appended to `sys.path` rather than inserting them from the beginning.

**Comment by GrahamDumpleton** [comment]
Am aware of the issue and was actually looking at it yesterday and wondering why I had it that way. I did uncover a bug in mod_wsgi which I fixed yesterday which was causing it to crash when both python-path and home options were supplied to WSGIDaemonProcess directive at the same time. I am not sure if I hit the problem and in my haste did it the alternative way until I was able to investigate the issue.

Anyway, your proposed solution doesn't necessarily go far enough. When simply adding extra paths into sys.path, it isn't the same as if you had set PYTHONPATH. This is because .pth files in a directory are not evaluated into order to work out any further paths that also need to be added for eggs or other variants of installed packages which use a subdirectory in the actual specified directory.

The better way of adding the paths is for the python-path option be added to WSGIDaemonProcess directive in the generated Apache configuration. This is because that option to WSGIDaemonProcess will handle .pth files and reorder the path so that any new directories are added first. I already had the code to change it to that sitting in my current checkout from yesterday. I wanted to think some more about the implications of the change and thus why I haven't committed it yet.

Part of the change to make things easier in mod_wsgi-express was to change things so that python-path being set to an empty string did nothing. In the past it would add an empty string to sys.path. This may not sound important but it is, as an empty string in sys.path has a special meaning and causes Python to look in the current working directory. I don't believe anyone would be relying on being able to do this and in practice the full path should be supplied anyway, but wanted to think through and make sure I wasn't making a change which would cause other issues.

So you should get a resolution on this pretty quick given I was already looking at it.

The alternate workaround right now is as explained in the help for --python-path option.

```
  --python-path DIRECTORY-PATH
                        Specify the path to any additional directory that
                        should be added to the Python module search path. Note
                        that these directories will not be processed for
                        '.pth' files. If processing of '.pth' files is
                        required, set the 'PYTHONPATH' environment variable in
                        a script specified by the '--envvars-script' option.
```

That is, create an envvars file and add to it:

```
    PYTHONPATH=/some/path
    export
```

and then use the option:

```
    --envvars-script envvars
```

As I said before, I am not sure why exactly I wasn't using the python-path option to WSGIDaemonProcess and instead was leaning towards requiring an envvars file to be used like that.

**Comment by GrahamDumpleton** [comment]
Can you try with:

* https://github.com/GrahamDumpleton/mod_wsgi/archive/develop.zip

and see if that now behaves as you expect.

You can use pip to install as:

```
pip install -U https://github.com/GrahamDumpleton/mod_wsgi/archive/develop.zip
```

**Comment by GrahamDumpleton** [comment]
The described change has been released in 4.4.10. Closing this issue. If still have issues then followup.

**Comment by tkdchen** [comment]
Hi, it's a quite a long time since reported this issue. Sorry for not reply in time at that time. Recently, I recalled this issue because I started to use mod_wsgi-express to run a WSGI app again, and I found that what I reported still happens. I installed 4.5.1 of mod-wsgi via pip. Both workaround you mentioned above and even just settings PYTHONPATH in the command line works.

**Comment by GrahamDumpleton** [comment]
I am going to ask a question I should have asked back then. Why should it override system paths?

First up, one should never chain Python virtual environments off a system wide Python. That causes so many problems it is not worth the trouble. You can see a variation of that issue explained in:

* http://blog.dscpl.com.au/2016/01/python-virtual-environments-and-docker.html

That is, per user installs cause similar problems.

As long as you do that, I can't see any good reason to add extra directories first. This will only hide mistakes such as incorrectly naming a module or package with the same name as one in the Python standard library.

So what is the use case you have that requires the paths be added first?

**Comment by tkdchen** [comment]
@GrahamDumpleton I think my followup misunderstood you :) As of writing comment, I only want to tell you this issue still happens as I encountered it again and you gave me a link above to confirm, before.

Why I use `--python-path` is because I read the option's description, "Specify the path to any additional directory that should be added to the Python module search path.". IMHO, I don't understand why specifying `--python-path` does not work just as the description tells, and don't know in which cases `--python-path` is enough and no need to specify PYTHONPATH or use `--envvars-scripts`. I tried to find answer from documentation and by searching in Google, but so far, didn't get an answer.

**Comment by GrahamDumpleton** [comment]
I am unclear on what you are saying is the problem. No point referring to the original report because when that occurred you got me when I had introduced a bug, which was fixed shortly after.

You are going to have to state what you are supplying as value to ``--python-path``, what you expect to see get added to ``sys.path`` within the process and where it appears in ``sys.path``. If it is getting added to the end of ``sys.path`` and you are expecting it to be added to the start, please explain why you need it to be at the start. Also explain how what is being done is causing a problem for your program.

Do note that a value you give it may not actually be added if the path is incorrect, either not existing, or there being no permissions to access it. The addition of a single path may also result in multiple directories being added to ``sys.path`` if the directory contain files with a ``.pth`` extension. These special directories may end up being added at the start where as the original directory may get added to the end. This simply mirrors what Python does for directories listed in ``.pth`` files.

**Comment by tkdchen** [comment]
> You are going to have to state what you are supplying as value to --python-path, what you expect to see get added to sys.path within the process and where it appears in sys.path. If it is getting added to the end of sys.path and you are expecting it to be added to the start, please explain why you need it to be at the start. Also explain how what is being done is causing a problem for your program.

Let's ignore this and the original request of my patch, because currently I'm not going to request a PR of my patch. :)

> Do note that a value you give it may not actually be added if the path is incorrect, either not existing, or there being no permissions to access it. The addition of a single path may also result in multiple directories being added to sys.path if the directory contain files with a .pth extension. These special directories may end up being added at the start where as the original directory may get added to the end. This simply mirrors what Python does for directories listed in .pth files.

I can confirm that the path passed to `--python-path` exists and accessible by the app. My case here is, I'm using mod_wsgi-express to run kojihub that is a typical WSGI application. I can run it with this command

```
PYTHONPATH=/home/vagrant/koji \
mod_wsgi-express start-server hub/kojixmlrpc.py --host 0.0.0.0 --port 8000 \
    --debug-mode \
    --working-rectory /home/vagrant/koji/hub/ \
    --setenv koji.hub.ConfigFile /home/vagrant/koji-dev-files/hub.conf
```
but, this one fails, 
```
mod_wsgi-express start-server hub/kojixmlrpc.py --host 0.0.0.0 --port 8000 \
    --debug-mode \
    --working-rectory /home/vagrant/koji/hub/ \
    --python-path /home/vagrant/koji \
    --setenv koji.hub.ConfigFile /home/vagrant/koji-dev-files/hub.conf
```
with error
```
[Sun Apr 10 10:57:42.493781 2016] [wsgi:error] [pid 2503:tid 139823722698816] Traceback (most recent call last):
[Sun Apr 10 10:57:42.493796 2016] [wsgi:error] [pid 2503:tid 139823722698816]   File "/tmp/mod_wsgi-0.0.0.0:8000:1000/handler.wsgi", line 94, in <module>
[Sun Apr 10 10:57:42.493864 2016] [wsgi:error] [pid 2503:tid 139823722698816]     recorder_directory=recorder_directory)
[Sun Apr 10 10:57:42.493884 2016] [wsgi:error] [pid 2503:tid 139823722698816]   File "/home/vagrant/koji-env/lib/python2.7/site-packages/mod_wsgi/server/__init__.py", line 1286, in __init__
[Sun Apr 10 10:57:42.494389 2016] [wsgi:error] [pid 2503:tid 139823722698816]     exec(code, self.module.__dict__)
[Sun Apr 10 10:57:42.494404 2016] [wsgi:error] [pid 2503:tid 139823722698816]   File "/home/vagrant/koji/hub/kojixmlrpc.py", line 33, in <module>
[Sun Apr 10 10:57:42.494579 2016] [wsgi:error] [pid 2503:tid 139823722698816]     from koji.server import WSGIWrapper
[Sun Apr 10 10:57:42.494600 2016] [wsgi:error] [pid 2503:tid 139823722698816] ImportError: No module named koji.server
```
where `koji/server.py` is under `/home/vagrant/koji/`

**Comment by GrahamDumpleton** [comment]
That is going to be an issue with ``--debug-mode`` flag as that isn't using daemon mode but embedded mode.

What happens if you aren't using the ``--debug-mode`` flag? Does it work for both ways then?

**Comment by GrahamDumpleton** [comment]
Try:

```
pip install -U https://github.com/GrahamDumpleton/mod_wsgi/archive/develop.zip
```

and see if that version works as you expect.

**Comment by tkdchen** [comment]
> and see if that version works as you expect.

Run
```
PYTHONPATH=/home/vagrant/koji \
mod_wsgi-express start-server hub/kojixmlrpc.py --host 0.0.0.0 --port 8000 \
    --debug-mode \
    --working-directory /home/vagrant/koji/hub/ \
    --setenv koji.hub.ConfigFile /home/vagrant/koji-dev-files/hub.conf
```
got this error
```
AH00526: Syntax error on line 260 of /tmp/mod_wsgi-0.0.0.0:8000:1000/httpd.conf:
WSGIPythonPath takes one argument, Python module search path.
```

and the other command works well.

**Comment by tkdchen** [comment]
> What happens if you aren't using the --debug-mode flag? Does it work for both ways then?

Yes. Without --debug-mode, both ways work.

**Comment by GrahamDumpleton** [comment]
Try again. Apache doesn't like an empty argument to the directive even when quoted.

**Comment by tkdchen** [comment]
> Try again. Apache doesn't like an empty argument to the directive even when quoted.

Both work now :+1:  Thanks. :)

**Comment by GrahamDumpleton** [comment]
Shipped in 4.5.2.

**Comment by tkdchen** [comment]
Great. Thank you very much.
